feat: add MMT lexer

ComFreek opened the pull request:
This adds a lexer for "MMT surface syntax", a language used within the [MMT project](https://uniformal.github.io/) for knowledge representation. It's some pretty mature project, albeit in a somewhat niche community of computer scientists.

Note that this lexer has been autogenerated from the [mmtpygments lexer](https://github.com/ComFreek/mmtpygments) (also by me) via the script [`pygments_to_rouge.py`](https://github.com/ComFreek/mmtpygments/blob/master/mmtpygments/pygments_to_rouge.py), hence the Rouge lexer is bound to be less idiomatic than would be possible when writing it manually.

BTW, is there general interest in such a conversion script? It works quite well for converting Pygments regex lexers to Rouge regex lexers: both, the Pygments *and* the Rouge lexer have been [tested on 14k LOC of MMT code](https://comfreek.github.io/mmtpygments/mmtpygments/test/index.html).

ComFreek committed:
feat: add MMT lexer

lambdaTotoro commented:
This would be lovely to have! Good job!

pyrmont committed:
Reformat rules

pyrmont committed:
Change tabs to spaces

pyrmont committed:
Use Unicode escape sequences

pyrmont committed:
Use goto rather than push when swapping states

pyrmont committed:
Remove error lexing

pyrmont committed:
Simplify rules in :root state

pyrmont committed:
Simplify rules in :theoryHeader state

pyrmont committed:
Simplify rules in :viewHeader state

pyrmont committed:
Simplify rules in :diagramHeader state

pyrmont committed:
Simplify rules in :moduleBody state

pyrmont committed:
Simplify rules in :constantDeclaration state

pyrmont committed:
Simplify rules in :notationExpression state

pyrmont committed:
Clarify rules in :structuralFeatureHeader state

pyrmont committed:
Fix typo in :moduleDefines state

pyrmont committed:
Reorder states

pyrmont committed:
Remove unused states

pyrmont commented on lib/rouge/lexers/mmt.rb:
```diff
@@ -0,0 +1,254 @@
+
+      state :moduleBody do
+        rule %r/\s+/im, Text::Whitespace
+        rule %r/\/T .*?(\u2759|\u275a)/im, Comment::Multiline
+        rule %r/\/\/.*?(\u2759|\u275a)/im, Comment::Multiline
```
These rules are slightly different to the multiline comment rules in the `:root` state. Was that intentional?

ComFreek replied:
Yes, multiline comments at root are delimited by "module delimiteres" (❚) while multiline comments in modules are delimited by "declaration delimiters" (❙) *or* "module delimiters" (❚).

pyrmont replied:
Got it!

pyrmont commented on lib/rouge/lexers/mmt.rb:
```diff
@@ -0,0 +1,254 @@
+          groups Punctuation, Name::Variable
+        end
+
+        rule(//) { goto :moduleDefines }
```
I'm not sure if Pygments supports this feature but Rouge allows you to essentially 'move' from one state to another using `goto` (really it pops the current state and then pushes the target state onto the stack). This is usually preferable because it means that subsequent states don't need to call `pop!` multiple times (which can lead to errors if you pop more states than are on the stack).

Oh, and as you may have guessed `rule (//)` will match anything without advancing the current character at all. This is a way to ensure that the lexer will move from state to another.

ComFreek replied:
Actually, I don't know whether Pygments supports moving without pushing the current state. 

> This is usually preferable because it means that subsequent states don't need to call pop! multiple times

But then, they can never return to the original state, can they? That is without knowing the original state being hardcoded in the general design.

In the MMT language theories (which are modules) can be highly nested ([see in sample here](https://github.com/rouge-ruby/rouge/blob/eedd87a005f8283866651b6317be739bcd812457/spec/visual/samples/mmt)). Hence, I designed the lexer to make heavy used of pushing and popping. I wouldn't know how to implement it otherwise.

pyrmont replied:
> But then, they can never return to the original state, can they? That is without knowing the original state being hardcoded in the general design.

That's correct, but the situations where I used calls to `goto` were to states that did not return to the immediately previous state. Rather they returned to earlier states (which they accomplished by calling `pop!` with a number greater than 1). Or at least that was my intent. Did I screw that up somewhere?

pyrmont commented on lib/rouge/lexers/mmt.rb:
```diff
@@ -0,0 +1,254 @@
+
+      state :root do
+        rule %r/\s+/im, Text::Whitespace
+        rule %r/\/T .*?\u275a/im, Comment::Multiline
```
So `/T ` can begin a multiline comment?

ComFreek replied:
Yes, there are multiple ways to enter comments in the MMT language: `// ... `, `/T ...`. They do have different meanings, but that's not important for the syntax highlighter.

pyrmont replied:
Understood!

pyrmont reviewed (changes requested):
@ComFreek Thanks for the submission! And a Pygments to Rouge translator is pretty cool! I've left notes in the code at particular places but wanted to make some high-level comments as well:

1. I have made adjustments to the file to make it consistent with the style used in Rouge. I tried to make my updated commits descriptive so hopefully it's easy to review. If you have any changes to make, there's no need to do any force pushing. This PR will be squashed when it's merged into master and I find having the complete history of work in the PR helpful for future reference.

2. I've also simplified the rules in many states to remove explicit error states. It's a non-goal for Rouge to have lexers check syntax and since the lexer was longish as it was, I felt removing these made the lexer easier to review and maintain.

3. I've replaced the Unicode graphemes in the source code with Unicode escape sequences. Ruby can handle these graphemes without issue but I think it can be difficult for debugging and maintenance (especially in text editors that don't provide easy ways to insert these graphemes). I didn't think this was an issue in the visual sample so I've left them there as-is.

Hopefully it's all clear, but let me know if anything doesn't seem to make sense.

ComFreek commented on lib/rouge/lexers/mmt.rb:
```diff
@@ -74,10 +74,17 @@ class MMT < RegexLexer
+          groups Punctuation, Name::Variable
+        end
+
+        rule(//) { goto :moduleDefiniens }
```
@pyrmont I don't see how these changes are semantics-preserving. If `theoryHeader` is entered and the text at hand matches none of the rules except the last, we will go to `moduleDefiniens`. Overall, the specific structure *required* by the MMT language for theory headers is not enforced anymore.

pyrmont replied:
I should have been clearer about these changes. The fact that the specific structure required by MMT is not enforced is what I mean when I say that syntax correctness is a non-goal. We generally don't recommend that lexers include this logic if doing so complicates the lexer's rules as it creates space for bugs in the long regular expressions that tend to be required to do this.

If the changes cause syntactically _correct_ code to be lexed incorrectly then this is a mistake on my part. I didn't see any differences in the visual sample you included in the original PR. Were you concerned that that would be the result here?

ComFreek replied:
> We generally don't recommend that lexers include this logic if doing so complicates the lexer's rules

For me, using a state for a single purpose is easier than using a state for multiple purposes ("substates") as is the case after this change.
But I see where your point regarding longish regular expressions is coming from.

> If the changes cause syntactically correct code to be lexed incorrectly

I am not sure if this is the case, probably not, but I would need to test this on my large test base. (will do the next days)

ComFreek commented:
Thanks for your comments!

1. > I tried to make my updated commits descriptive so hopefully it's easy to review.

   If I may, I'd like to suggest commit messages other than "simplify X" for commits simply changing whitespace/indentation and not semantics at all :smile:

2. > I've also simplified the rules in many states to remove explicit error states. It's a non-goal for Rouge to have lexers check syntax and since the lexer was longish as it was, I felt removing these made the lexer easier to review and maintain.

   Alright. Actually, I am not sure whether these error states ever worked as intended in the Pygments one, either.

3. > I've replaced the Unicode graphemes in the source code with Unicode escape sequences. Ruby can handle these graphemes without issue but I think it can be difficult for debugging and maintenance (especially in text editors that don't provide easy ways to insert these graphemes).

   By contrast, *now* I find it hard to debug and maintain your changes/the lexer. I don't know the Unicode escape sequences by heart; seeing the Unicode characters is much easier. I've coded the Pygments lexer fully in VisualStudio Code which doesn't have a way to insert them, either, and copy-pasting worked fine.

   (PS: In the IDEs for the MMT language, there are easy shortcuts to insert those. And I've recently developed [MMT OnlineTools](https://comfreek.github.io/mmteditor/) as yet another shortcut when you don't have any IDE installed.)

4. In some commits, you cut out parts of regexes, reordered them, and split them. Back then, I started the lexer from scratch and only complicated parts if really needed. This was a non-trivial task to do and spanned multiple days until [14k LOC of MMT code](https://comfreek.github.io/mmtpygments/mmtpygments/test/index.html) highlighting correctly.

   In one case, I left a GitHub comment at the respective lines stating that the changes really changed/broke (?) the semantics.

   In general, as the author of the lexer itself, I'd be cautious to do any substantial changes in the rules myself without (a) rerunning of those 14k LOC and (b) understanding the relevant part of the MMT language myself. (I learned lots of new parts of the MMT language by developing the lexer!)

Overall, I'd like to further discuss point 4. I don't care too much about 3, though, as I could just instruct my Pygments2Rouge converter to output the Unicode sequences instead of the graphemes. But point 4 (i) breaks semantics in some (possibly unexpected/yet unknown) ways and (ii) I would need to merge those changes back into the Pygments one.

TODO list for the Pygments lexer:

- [ ] remove/comment out error states
- [ ] remove unused states (expect{OD|DD|MD)}

TODO list for the Pygments2Rouge converter:

- [ ] possibly convert graphemes (depending on what @pyrmont comments back)
- [ ] use spaces instead of tabs
- [ ] use `, ` with space for separating lists

PS: I now also have a Pygments2CodeMirror converter running successfully :smile: The [MMT OnlineTools](https://comfreek.github.io/mmteditor/) linked above exactly use my Pygments lexer as converted by that!

ComFreek reviewed (commented):
see comments in comment thread

pyrmont commented:
> By contrast, now I find it hard to debug and maintain your changes/the lexer. I don't know the Unicode escape sequences by heart; seeing the Unicode characters is much easier. I've coded the Pygments lexer fully in VisualStudio Code which doesn't have a way to insert them, either, and copy-pasting worked fine.

I guess different strokes for different folks. I was editing these files in Vim and it was not easy to tell the difference between the various vertical lines (I didn't realise until using the Unicode escape sequences that the thinnest vertical bar was not `|`). I realise a person is not likely to know the escape sequences by heart but I guess I greatly prefer something that involves ASCII characters rather than unusual graphemes.

Do you have an opinion on this, @jneen? I don't think we have other lexers where this issue has come up before so I'm not sure how we should approach it.

> In one case, I left a GitHub comment at the respective lines stating that the changes really changed/broke (?) the semantics.

Do you mean you left a comment in respect to my changes? I'm sorry, I didn't see it. Would you be able to link to it here?

> In general, as the author of the lexer itself, I'd be cautious to do any substantial changes in the rules myself without (a) rerunning on those 14k LOC and (b) understanding the relevant part of the MMT language myself. (I learned lots of new parts of the MMT language by developing the lexer!).
>
> I would like to kindly ask you (a) whether you tested your changes, (b) why the changes were needed, and (c) whether the reasons for (b) really outweigh the downside of breaking semantics.

I guess my view is that lexers that are submitted need to be maintainable by persons other than the original author. If lexer rules are complex and difficult to analyse, that's an argument strongly in favour of simplification in my view.

How do you feel about this, @jneen? The original submission from @ComFreek was clearly the result of a significant investment of time and so I'm sympathetic to arguments to keep them largely as they were. My concern was extremely long and complicated regular expressions. However, perhaps a better solution would have been to use non-whitespace preserving regular expressions (i.e. using the `x` flag) that were easier to read rather than breaking the rules apart.

ComFreek commented:
> Do you mean you left a comment in respect to my changes? I'm sorry, I didn't see it. Would you be able to link to it here?

You already answered there :smile:

> I guess my view is that lexers that are submitted need to be maintainable by persons other than the original author.

Perhaps to give some perspective: there are <= 15 people on earth who understand the MMT language and ecosystem well enough to reason in-depth about it. All *all* of those belong to one research group and know each other personally. There are probably <= 5 people on earth who know all the syntax to be able to maintain a syntax highlighter. And probably only <= 3 (myself included) would invest considerable time in this. (You know, professors are busy.)

Hence, while I see your general point of maintenance being possible by third parties, it might not really apply to our case at hand. In my view, the lexer would even get *more* maintainable if it were to be kept as the auto-generated one from the Pygments lexer. (+ minor changes as tabs vs. spaces etc.)

ComFreek commented:
Regarding including comments in regexes (using the `x` flag): I do have [extensive comments in the Pygments lexer](https://github.com/ComFreek/mmtpygments/blob/59b9c0459ebc78e81cff360d71a9441f4aaa1678/mmtpygments/mmt_lexer.py#L125-L153). If there was any easy way to carry those over to the Rouge one, I'd happily do so. But I doubt there's a way.
